### ISSUE TITLE
Reward labeling is too slow

### ISSUE THREAD

**Issue by raojay7** [RLHFlow/Online-DPO-R1#9]
How can this process be accelerated?

**Comment by WeiXiongUST** [comment]
You can try any other annotation script actually. For instance, I update the ToRA script in https://github.com/WeiXiongUST/Building-Math-Agents-with-Multi-Turn-Iterative-Preference-Learning/tree/main/inference 

This usually takes only less than 2 minutes to label for each iteration. You need to prepare your dataset with the feature 'code', which is a list of answers (generated by VLLM). Note that you do not specify --data_name gsm8k as long as you have feature "gt" in your file.

**Comment by raojay7** [comment]
Thank you for your reply. Is it possible to merge directly into the code of this repository?

**Comment by WeiXiongUST** [comment]
sure, I will do this in next two days.

**Comment by WeiXiongUST** [comment]
Hi I have add an alternative code to label the data. You can call it by 
```sh
python -um trainning_data_annotation.annotate_data --dataset_name_or_path local.jsonl --output_dir out.jsonl
```

**Comment by WeiXiongUST** [comment]
You need additional dependence: regex, pebble timeout_decorator

**Comment by raojay7** [comment]
Thank you for your efforts.

**Comment by pspdada** [comment]
> Hi I have add an alternative code to label the data. You can call it by
> 
> python -um trainning_data_annotation.annotate_data --dataset_name_or_path local.jsonl --output_dir out.jsonl

I encountered some warnings when using this script. Do you have any ideas about whether it will give wrong results or how to solve it?

First
```
<string>:1: SyntaxWarning: 'set' object is not callable; perhaps you missed a comma?
<string>:1: SyntaxWarning: 'set' object is not callable; perhaps you missed a comma?
<string>:1: SyntaxWarning: 'set' object is not callable; perhaps you missed a comma?
```

Second:
```
/lib/python3.10/site-packages/sympy/core/evalf.py:1749: SymPyDeprecationWarning: 

The string fallback in sympify() is deprecated.

To explicitly convert the string form of an object, use
sympify(str(obj)). To add define sympify behavior on custom
objects, use sympy.core.sympify.converter or define obj._sympy_
(see the sympify() docstring).

sympify() performed the string fallback resulting in the following string:

'<function solve_univariate_inequality at 0x71cbaa7d53f0>'

See https://docs.sympy.org/latest/explanation/active-deprecations.html#deprecated-sympify-string-fallback
for details.

This has been deprecated since SymPy version 1.6. It
will be removed in a future version of SymPy.

  return sympify(x, rational=True).evalf(n, **options)
```

**Comment by WeiXiongUST** [comment]
This is very common when you use sympy to label data. I encounter these for ToRA, Qwen, and many other scripts. Make sure sympy==1.12 though

**Comment by pspdada** [comment]
It seems that this script does not handle all cases well. For example:

```python
a = "\nThe calculation confirms that 20 centimeters on the map represents 120 kilometers in real life. So the final answer is:\n\n\\[\n\\boxed{120}\n\\]"
b = "120 \\text{ kilometers}"
res = get_batch_scores([[a]], [b])
print(res)
```

When the ground truth (gt) contains additional information (such as units), the result will be `false`.

**Comment by pspdada** [comment]
I have discovered another issue while using this script. For a sample size of 20k, the majority of the samples are easy to evaluate and can be completed in a very short time, while a small portion takes a very long time, accounting for the vast majority of the total time. Do you have any thoughts on this? Are those samples that take a long time more likely to have issues, potentially affecting the quality of the training dataset?
```
Execute: 100%|████████████| 20000/20000 [00:04<00:00, 4180.28it/s]
Evaluate:  86%|██████████▎ | 17255/20000 [00:13<00:02, 1249.87it/s]
Evaluate: 32286it [00:23, 2924.24it/s]                                                                                                                                           
Evaluate: 39937it [00:34, 1308.37it/s]('Task timeout', 10)
Evaluate: 100517it [01:15, 1253.68it/s]
Evaluate: 117231it [01:31, 1286.97it/s]numerator
therangeofvaluesfortherealnumber\(a\)is\([-4,+\infty)\).
('Task timeout', 10)
Evaluate: 127525it [01:40, 1426.63it/s]('Task timeout', 10)
Evaluate: 152036it [01:52, 2090.89it/s]
Evaluate: 160000it [02:03, 1297.09it/s]
Time: 147.34145641326904
```

**Comment by pspdada** [comment]
Incorrect Parsing of `^` and Reliance on Relative Tolerance Leading to False Positives

There is an issue with the current implementation of parsing and comparison logic that leads to incorrect results when evaluating whether two mathematical expressions are equivalent. Specifically, the problem arises due to two main issues:

**Incorrect Parsing of `^` as Bitwise XOR Instead of Exponentiation:**

   When using `parse_expr` from `sympy.parsing.sympy_parser`, the default behavior does not interpret `^` as a power operator. Instead, it follows Python's semantics and treats `^` as a bitwise XOR operator. This leads to incorrect results when parsing expressions involving exponents.

   For example:
   ```python
   from sympy.parsing.sympy_parser import parse_expr

   print(parse_expr("2^3"))   # Output: 1 (incorrect, should be 8)
   print(parse_expr("2^10"))  # Output: 8 (incorrect, should be 1024)
   print(parse_expr("2^20"))  # Output: 22 (incorrect, should be 1048576)
   ```

   These results clearly demonstrate that `^` is being parsed incorrectly as a bitwise XOR operator, which is not the intended behavior for mathematical expressions.

**Reliance on Relative Tolerance Without Absolute Tolerance:**

   In addition to the parsing issue, the use of `math.isclose` with only a relative tolerance (`rel_tol`) can lead to false positives when comparing large numbers. For instance:
   ```python
   from math import isclose

   print(isclose(2010, 2009, rel_tol=1e-3))  # Output: True
   ```

   The above result is `True` because the relative difference between `2010` and `2009` is small compared to their magnitudes. However, this can cause problems when comparing expressions like `4^{2008}` and `2^{2009}`. If these expressions are incorrectly parsed using bitwise XOR, the resulting values might appear close in relative terms, leading to an incorrect match.

   For example:
   ```python
   a = "box the final answer:\n\n\\[\n\\boxed{4^{2008}}\n\\]"
   b = "2^{2009}"
   res = get_batch_scores([[a]], [b])
   print(res)  # Should be False, but returns True
   ```

   The root cause of this issue is that the expressions are parsed incorrectly (due to the `^` issue), and the subsequent comparison relies solely on relative tolerance, which fails to account for absolute differences.


**Proposed Solutions:**

1. **Correctly Parse `^` as Exponentiation:**
   To ensure that `^` is interpreted as a power operator, the `parse_expr` function should be used with the `convert_xor` transformation rule. For example:
   ```python
   from sympy.parsing.sympy_parser import parse_expr, standard_transformations, convert_xor

   transformations = (standard_transformations + (convert_xor,))
   print(parse_expr("2^3", transformations=transformations))   # Output: 8
   print(parse_expr("2^10", transformations=transformations))  # Output: 1024
   print(parse_expr("2^20", transformations=transformations))  # Output: 1048576
   ```

2. **Use Both Relative and Absolute Tolerances:**
   When comparing numerical values, both relative and absolute tolerances should be considered to avoid false positives. For example:
   ```python
   from math import isclose

   print(isclose(2010, 2009, rel_tol=1e-3, abs_tol=1e-9))  # Output: False
   ```

   By setting an appropriate `abs_tol`, we can ensure that small differences (e.g., due to incorrect parsing) do not lead to incorrect matches.
